hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java

```diff
@@ -37,152 +40,294 @@
+        }
+      }
     }
-    catch (SQLException e) {
-      logger.error("Cannot open connection", e);
-      exceptionOnConnect = e;
+
+    for (String key : removeKeySet) {
+      propertiesMap.remove(key);
     }
+
+    logger.debug("propertiesMap: {}", propertiesMap);
   }
 
   @Override
   public void close() {
     try {
-      if (jdbcConnection != null) {
-        jdbcConnection.close();
+      for (Statement statement : paragraphIdStatementMap.values()) {
+        statement.close();
       }
+      paragraphIdStatementMap.clear();
+
+      for (Connection connection : keyConnectionMap.values()) {
+        connection.close();
+      }
+      keyConnectionMap.clear();
+    } catch (SQLException e) {
+      logger.error("Error while closing...", e);
     }
-    catch (SQLException e) {
-      logger.error("Cannot close connection", e);
+  }
+
+  public Connection getConnection(String propertyKey) throws ClassNotFoundException, SQLException {
+    Connection connection = null;
+    if (keyConnectionMap.containsKey(propertyKey)) {
+      connection = keyConnectionMap.get(propertyKey);
+      if (connection.isClosed() || !connection.isValid(10)) {
+        connection.close();
+        connection = null;
+        keyConnectionMap.remove(propertyKey);
+      }
     }
-    finally {
-      jdbcConnection = null;
-      exceptionOnConnect = null;
+    if (null == connection) {
+      Properties properties = propertiesMap.get(propertyKey);
+      Class.forName(properties.getProperty(DRIVER_KEY));
+      String url = properties.getProperty(URL_KEY);
+      String user = properties.getProperty(USER_KEY);
+      String password = properties.getProperty(PASSWORD_KEY);
+      if (null != user && null != password) {
+        connection = DriverManager.getConnection(url, user, password);
+      } else {
+        connection = DriverManager.getConnection(url, properties);
+      }
+      keyConnectionMap.put(propertyKey, connection);
     }
+    return connection;
   }
 
-  Statement currentStatement;
-  private InterpreterResult executeSql(String sql) {
-    try {
-      if (exceptionOnConnect != null) {
-        return new InterpreterResult(Code.ERROR, exceptionOnConnect.getMessage());
+  public Statement getStatement(String propertyKey, String paragraphId)
+      throws SQLException, ClassNotFoundException {
+    Statement statement = null;
+    if (paragraphIdStatementMap.containsKey(paragraphId)) {
+      statement = paragraphIdStatementMap.get(paragraphId);
+      if (statement.isClosed()) {
+        statement = null;
+        paragraphIdStatementMap.remove(paragraphId);
       }
-      currentStatement = jdbcConnection.createStatement();
+    }
+    if (null == statement) {
+      statement = getConnection(propertyKey).createStatement();
+      paragraphIdStatementMap.put(paragraphId, statement);
+    }
+    return statement;
+  }
+
+  public InterpreterResult executeSql(String propertyKey,
+                                      String sql,
```

Comment:
What do you think, does horizontal alignment makes it more readable or just makes it longer?

According to the styleguide Zeppelin project uses - it is never required and has some cons https://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment